File: apps/api/src/schemas/job.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class AnalysisJobOut(BaseModel):
    id: str
    video_id: str
    status: Literal["queued", "processing", "done", "failed"]
    current_stage: str | None
    progress_pct: int
    error_message: str | None
    estimated_duration_seconds: int | None
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime


class DetectedPlayerOut(BaseModel):
    id: str
    video_id: str
    track_id: int
    label: str
    ocr_jersey_number: int | None
    ocr_confidence: float | None
    thumbnail_path: str | None
    athlete_id: str | None
    linked_at: datetime | None


class PlayerLinkRequest(BaseModel):
    athlete_id: str


class PlayerMatchStatsOut(BaseModel):
    id: str
    video_id: str
    match_id: str | None
    detected_player_id: str
    athlete_id: str | None
    touches: int
    passes_completed: int
    passes_attempted: int
    dribbles_successful: int
    dribbles_attempted: int
    fouls_committed: int
    fouls_suffered: int
    shots: int
    distance_km: float | None
    avg_speed_kmh: float | None
    max_speed_kmh: float | None
    heatmap: list[list[int]] | None


class ClipOut(BaseModel):
    id: str
    event_id: str
    storage_path: str
    thumbnail_path: str | None
    duration_seconds: int | None
    playback_url: str | None = None


class EventOut(BaseModel):
    id: str
    video_id: str
    match_id: str | None
    event_type: Literal[
        "passe", "drible", "desarme", "falta", "finalizacao", "triangulacao", "jogada_ensaiada"
    ]
    start_time_ms: int
    end_time_ms: int
    primary_player_id: str | None
    secondary_player_id: str | None
    confidence: float | None
    clips: list[ClipOut] = []

File: packages/shared-types/src/job.ts
export type AnalysisJobStatus = "queued" | "processing" | "done" | "failed";

export interface AnalysisJob {
  id: string;
  video_id: string;
  status: AnalysisJobStatus;
  current_stage: string | null;
  progress_pct: number;
  error_message: string | null;
  estimated_duration_seconds: number | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
}

export interface DetectedPlayer {
  id: string;
  video_id: string;
  track_id: number;
  label: string;
  ocr_jersey_number: number | null;
  ocr_confidence: number | null;
  thumbnail_path: string | null;
  athlete_id: string | null;
  linked_at: string | null;
}

export interface PlayerMatchStats {
  id: string;
  video_id: string;
  match_id: string | null;
  detected_player_id: string;
  athlete_id: string | null;
  touches: number;
  passes_completed: number;
  passes_attempted: number;
  dribbles_successful: number;
  dribbles_attempted: number;
  fouls_committed: number;
  fouls_suffered: number;
  shots: number;
  distance_km: number | null;
  avg_speed_kmh: number | null;
  max_speed_kmh: number | null;
  heatmap: number[][] | null;
}

export interface ReferenceFrame {
  id: string;
  video_id: string;
  storage_path: string;
  image_url: string;
  timestamp_ms: number;
  frame_width: number;
  frame_height: number;
  created_at: string;
}

export interface PlaybackUrl {
  video_id: string;
  url: string;
}

export interface SelectionHint {
  id: string;
  video_id: string;
  athlete_id: string;
  timestamp_ms: number;
  bbox_x1: number;
  bbox_y1: number;
  bbox_x2: number;
  bbox_y2: number;
  frame_width: number;
  frame_height: number;
  created_by: string;
  created_at: string;
}

export type EventType =
  | "passe"
  | "drible"
  | "desarme"
  | "falta"
  | "finalizacao"
  | "triangulacao"
  | "jogada_ensaiada";

export interface MatchEvent {
  id: string;
  video_id: string;
  match_id: string | null;
  event_type: EventType;
  start_time_ms: number;
  end_time_ms: number;
  primary_player_id: string | null;
  secondary_player_id: string | null;
  confidence: number | null;
}

export interface Clip {
  id: string;
  event_id: string;
  storage_path: string;
  thumbnail_path: string | null;
  duration_seconds: number | null;
}
